examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/ntp.c

```diff
@@ -0,0 +1,301 @@
+
+    printf("\r\n>>Parsing SNTP response...");
+    memcpy(&current_ntp_time, response+40, 4); //timestamp starts at byte 40
+    current_ntp_time = sntp_ntohl(current_ntp_time); //Convert endianness
+    current_epoch_time = current_ntp_time - 2208988800; //Remove NTP offset
+    return current_epoch_time;
+}
+
+/**
+   * @brief  Create new UDP socket to SNTP server, send SNTP request, await
+   *         response for up to SNTP_TIMEOUT_MS, close socket and return
+   * @param  - sock_id is a pre-allocated uint8_t pointer used to filter the WiFi
+   *         API callbacks. It will be set to the socket ID returned by connect()
+   *         - epoch_time will be set to the current epoch time as returned by the
+   *         SNTP server
+   * @retval sntp_status_t: 0 means SNTP_SUCCESS, <0 means something failed.
+   *         See ntp.h to handle different failure reasons in different ways
+   */
+sntp_status_t sntp_get_datetime( uint8_t* sock_id, int32_t* epoch_time )
+{
+    sntp_status_t retval = SNTP_SUCCESS; //Returned by this function
+    WiFi_Status_t wifi_retval = WiFi_MODULE_SUCCESS;
+
+    /* Create socket */
+    printf("\r\n>>Getting date and time from SNTP server...");
+    wifi_retval = sntp_connect(SNTP_SERVER, SNTP_PORT, sock_id);
+    if( wifi_retval != WiFi_MODULE_SUCCESS )
+    {
+        printf("\r\n\tSNTP socket connection [FAIL] Retval: %d", wifi_retval);
+        retval = SNTP_CONNECTION_FAILURE;
+        goto terminate;
+    }
+    else if(*sock_id < 0)
+    {
+        printf("\r\n\tUDP Connection [FAIL] Socket ID<0: %d", *sock_id);
+        retval = SNTP_CONNECTION_FAILURE;
+        goto terminate;
+    }
+    printf("\r\n\tUDP Connection [OK] Assigned socket ID: %d", *sock_id);
+
+    /* Send SNTP request */
+    wifi_retval = sntp_send_request(*sock_id);
+    if( wifi_retval != WiFi_MODULE_SUCCESS )
+    {
+        printf("\r\n\tSNTP send_request [FAIL] Retval: %d", wifi_retval);
+        retval = SNTP_REQUEST_FAILURE;
+        goto terminate;
+    }
+
+    /* Await SNTP Response */
+    printf("\r\n\tAwaiting server response");
+    if( sntp_await_response(*sock_id) < 0 )
+    {
+        printf("\r\n\tSNTP response [FAIL] Response timed out");
+        retval = SNTP_TIMEOUT;
+        goto terminate;
+    }
+
+    if( *sock_id != last_sntp_response->socket_id )
+    {
+        printf("\r\n\tSocket ID assertion [FAIL] Bug in ntp.c?");
+        retval = SNTP_INTERNAL_ERROR;
+        goto terminate;
+    }
+
+    /* Parse server response */
+    printf("\r\n\tParsing and converting SNTP response");
+    sntp_current_time = sntp_parse_response(last_sntp_response->response);
+    if( sntp_current_time < 0 )
+    {
+        printf("\r\n\tSNTP Parsing and conversion [FAIL]");
+        retval = SNTP_PARSER_ERROR;
+    }
+    else
+    {
+        printf("\r\n\tReceived epoch time: %ld", sntp_current_time);
+        retval = SNTP_SUCCESS;
+    }
+
+terminate:
+    /* free() SNTP Response */
+    if( NULL != last_sntp_response )
```

Comment:
Just a minor comment - you could move this logic of checking the NULLness into the implementation of the sntp_free_response - you can also pass the ```**sntp_response``` this way you can set it NULL internally as well. This will increase the readability of this section of the code.